### DISCUSSION TITLE
HassIO/Raspberry Gazpar custom component installation

### DISCUSSION THREAD

**Discussion by yvavasseur** [ssenart/PyGazpar#9] [General]
Bonjour,
Est-il possible d'installer PyGazpar sur Home Assistant OS ?
J'ai bien vu comment installer le custom_components, mais il requiert (ce qui est logique) cette librairie PyGazpar.
Il est possible de l'installer facilement sous HomeAssistant OS ? Merci d'avance :)

**Comment by ssenart** [comment]
L'installation du package PyGazpar est automatique à partir du custom_component correspondant. En effet, la dépendance est déclarée dans le fichier manifest.json, et Home Assistant va le télécharger et l'installer au démarrage.
Donc, il n'y a rien à faire d'autre que d'installer le custom_component gazpar disponible ici : 
[https://github.com/ssenart/ha-custom_components/tree/master/gazpar](https://github.com/ssenart/ha-custom_components/tree/master/gazpar)

**Comment by yvavasseur** [comment]
Merci de ta réponse !
Malheureusement, cela ne suffit visiblement pas, j'obtiens ce message d'erreur toujours : 
`Failed to query PyGazpar library with exception : Traceback (most recent call last): File "/usr/local/lib/python3.8/site-packages/selenium/webdriver/common/service.py", line 72, in start self.process = subprocess.Popen(cmd, env=self.env, File "/usr/local/lib/python3.8/subprocess.py", line 854, in __init__ self._execute_child(args, executable, preexec_fn, close_fds, File "/usr/local/lib/python3.8/subprocess.py", line 1702, in _execute_child raise child_exception_type(errno_num, err_msg, err_filename) FileNotFoundError: [Errno 2] No such file or directory: '/config/drivers/geckodriver' During handling of the above exception, another exception occurred: Traceback (most recent call last): File "/config/custom_components/gazpar/sensor.py", line 124, in update_gazpar_data client.update() File "/usr/local/lib/python3.8/site-packages/pygazpar/client.py", line 90, in update driver = webdriver.Firefox(executable_path=self.__firefox_webdriver_executable, firefox_profile=profile, options=options, service_log_path=geckodriverLogFile) File "/usr/local/lib/python3.8/site-packages/selenium/webdriver/firefox/webdriver.py", line 164, in __init__ self.service.start() File "/usr/local/lib/python3.8/site-packages/selenium/webdriver/common/service.py", line 81, in start raise WebDriverException( selenium.common.exceptions.WebDriverException: Message: 'geckodriver' executable needs to be in PATH.`
J'ai l'impression que c'est à cause de "gecko driver" ? faut-il que je l'installe manuellement ?
Merci d'avance ! 👍

**Comment by ssenart** [comment]
Ah, oui, en effet, il faut le binaire du geckodriver et une installation firefox compatible.

Et j'allais oublier le configuration HA qui va avec : 

- platform: gazpar
  username: !secret gazpar.username
  password: !secret gazpar.password
  webdriver: /config/drivers/geckodriver
  tmpdir: /config/tmp
  scan_interval: 01:00:00

**Comment by yvavasseur** [comment]
Merci! Le configuration j’avais déjà trouvé dans un précédent post.
installer le geckodriver depuis home assistant OS (sur raspberry pi) c’est possible ?
ainsi que Firefox ? Depuis une console ssh ?
Merci d’avance :)

**Comment by ssenart** [comment]
Firefox: oui, via une console Ssh, installe le paquet Firefox qui va bien.
Geckodriver: c'est un binaire unique: je le met dans un sous dossier du dossier de config de HA. Et, surtout, je m'assure du chemin correct dans le fichier de config de HA.

**Comment by yvavasseur** [comment]
Le problème est que je suis sur Home Assistant OS directement, et pas sur une raspbian classique. Je ne vois pas comment installer le paquet firefox, même en ssh les commandes apt install ne sont pas reconnues... :(

**Comment by ssenart** [comment]
Je n'utilise pas Home Assistant OS. C'est probablement une distribution Linux fonctionnant avec un autre gestionnaire de package que apt. Alpine Linux est souvent utilisé comme distribution légère dans ce genre de contexte (comme docker). Son gestionnaire de package est apk (https://wiki.alpinelinux.org/wiki/Alpine_Linux_package_management). Tu peux essayer ce dernier pour commencer.

**Comment by yvavasseur** [comment]
Merci de ton retour ! en effet, il s'agit bien du gestionnaire apk. Cependant, après de multiples essais, je pense que cela ne peut pas fonctionner car je fais tourner home assistant sur une Raspberry Pi, et malheureusement le Geckodriver n'est pas compatible avec les plateformes ARM (en tout cas pas depuis un certain temps)...

**Comment by ssenart** [comment]
As-tu essayer avec la version 0.23 : https://github.com/mozilla/geckodriver/releases/download/v0.23.0/geckodriver-v0.23.0-arm7hf.tar.gz
Et firefox >= 66.

**Comment by yvavasseur** [comment]
Oui j’ai essayé justement celle ci, et avec Firefox-esr 78.6.1, mais toujours des messages d’erreurs,  notamment :

FileNotFoundError: [Errno 2] No such file or directory: '/config/drivers/geckodriver' 

Alors que j’ai bien mis le fichier geckodriver dans le dossier drivers du dossier config. J’ai même modifié les droits du fichier pour le rendre exécutable, rien n’y fait...

Envoyé de mon iPad

**Comment by ssenart** [comment]
Ce n'est pas un problème de version car il n'arrive même pas à trouver le binaire. C'est peut-être un problème de droits.
Pour éliminer ce cas, peux-tu essayer la commande et relancer :
`# chmod 777 /config/drivers/geckodriver`
Cela donne tous les droits à tout le monde.

**Comment by yvavasseur** [comment]
Malheureusement j’avais essayé un 777 mais cela ne change rien...
Toujours les messages d’erreurs :

Failed to query PyGazpar library with exception : Traceback (most recent call last): File "/usr/local/lib/python3.8/site-packages/selenium/webdriver/common/service.py", line 72, in start self.process = subprocess.Popen(cmd, env=self.env, File "/usr/local/lib/python3.8/subprocess.py", line 854, in __init__ self._execute_child(args, executable, preexec_fn, close_fds, File "/usr/local/lib/python3.8/subprocess.py", line 1702, in _execute_child raise child_exception_type(errno_num, err_msg, err_filename) FileNotFoundError: [Errno 2] No such file or directory: '/config/drivers/geckodriver' During handling of the above exception, another exception occurred: Traceback (most recent call last): File "/config/custom_components/gazpar/sensor.py", line 124, in update_gazpar_data client.update() File "/usr/local/lib/python3.8/site-packages/pygazpar/client.py", line 90, in update driver = webdriver.Firefox(executable_path=self.__firefox_webdriver_executable, firefox_profile=profile, options=options, service_log_path=geckodriverLogFile) File "/usr/local/lib/python3.8/site-packages/selenium/webdriver/firefox/webdriver.py", line 164, in __init__ self.service.start() File "/usr/local/lib/python3.8/site-packages/selenium/webdriver/common/service.py", line 81, in start raise WebDriverException( selenium.common.exceptions.WebDriverException: Message: 'geckodriver' executable needs to be in PATH.

Sachant que le geckodriver est bien dans config/drivers/, et que voici le config.yaml :

sensor:
- platform: gazpar
    username: ***
    password: ***
    webdriver: /config/drivers/geckodriver
    tmpdir: /tmp
    scan_interval: 01:00:0

Vois tu d’où ça pourrait venir ?

Envoyé de mon iPad

**Comment by ssenart** [comment]
Ok, et même question sur le dossier parent /config/drivers. Mettre le maximum de droits pour tous.

Aussi, arrives-tu à exécuter manuellement ce binaire via la console SSH ?

**Comment by yvavasseur** [comment]
Oui, je lui ai mis 777 aussi dans le doute.

Mais je n’arrive pas à l’exécuter depuis la console. Si je fais un ls - /config/drivers/ il me trouve bien le fichier geckodriver avec tous les droits. 

Mais si j’essaie de l’exécuter /config/drivers/geckodriver il me dit no such file or directory... or il est bien là ! ...

**Comment by ssenart** [comment]
Ok, on y est ! Il y a une incohérence quelque part:
- la commande ls te remonte l'existence du fichier avec les droits en exécution.
- l'exécution du fichier échoue avec le message : No such file or directory.

Ca ressemble à une corruption du système de fichier. J'imagine que tu es sur une carte SD, et c'est fréquent avec ce genre de support.
Plusieurs choses à essayer dans l'ordre:
1. Ramener un nouveau binaire geckodriver mais le copier ailleurs (par exemple : /config). Placer les bons droits, et l'exécuter avec la commande:
`# geckodriver --version`
Si toujours le problème, passer aux étapes suivantes, sinon reconfigurer HA avec le nouveau chemin et le redémarrer.
2. Réparer le système de fichier (google t'aidera).
3. Changer la carte SD.

**Comment by yvavasseur** [comment]
Le problème ne vient pas de la carte SD. J'ai testé sur une nouvelle installation de HomeAssistant sur une carte SD neuve, toujours le même problème.

J'ai exploré plusieurs pistes : 

1. Quand je ne modifie pas les droits sur le geckodriver, j'obtiens ces erreurs ci sous HomeAssistant : 
`Failed to query PyGazpar library with exception : Traceback (most recent call last): File "/usr/local/lib/python3.8/site-packages/selenium/webdriver/common/service.py", line 72, in start self.process = subprocess.Popen(cmd, env=self.env, File "/usr/local/lib/python3.8/subprocess.py", line 854, in __init__ self._execute_child(args, executable, preexec_fn, close_fds, File "/usr/local/lib/python3.8/subprocess.py", line 1702, in _execute_child raise child_exception_type(errno_num, err_msg, err_filename) PermissionError: [Errno 13] Permission denied: '/config/drivers/geckodriver' During handling of the above exception, another exception occurred: Traceback (most recent call last): File "/config/custom_components/gazpar/sensor.py", line 124, in update_gazpar_data client.update() File "/usr/local/lib/python3.8/site-packages/pygazpar/client.py", line 90, in update driver = webdriver.Firefox(executable_path=self.__firefox_webdriver_executable, firefox_profile=profile, options=options, service_log_path=geckodriverLogFile) File "/usr/local/lib/python3.8/site-packages/selenium/webdriver/firefox/webdriver.py", line 164, in __init__ self.service.start() File "/usr/local/lib/python3.8/site-packages/selenium/webdriver/common/service.py", line 86, in start raise WebDriverException( selenium.common.exceptions.WebDriverException: Message: 'geckodriver' executable may have wrong permissions.
`
**Permission denied: '/config/drivers/geckodriver'** --> je donne les droits 777 sur le fichier, j'obtiens alors comme avant 

2. Avec les droits 777 et plusieurs versions du geckodriver différent : 
`Failed to query PyGazpar library with exception : Traceback (most recent call last): File "/usr/local/lib/python3.8/site-packages/selenium/webdriver/common/service.py", line 72, in start self.process = subprocess.Popen(cmd, env=self.env, File "/usr/local/lib/python3.8/subprocess.py", line 854, in __init__ self._execute_child(args, executable, preexec_fn, close_fds, File "/usr/local/lib/python3.8/subprocess.py", line 1702, in _execute_child raise child_exception_type(errno_num, err_msg, err_filename) FileNotFoundError: [Errno 2] No such file or directory: '/config/drivers/geckodriver' During handling of the above exception, another exception occurred: Traceback (most recent call last): File "/config/custom_components/gazpar/sensor.py", line 124, in update_gazpar_data client.update() File "/usr/local/lib/python3.8/site-packages/pygazpar/client.py", line 90, in update driver = webdriver.Firefox(executable_path=self.__firefox_webdriver_executable, firefox_profile=profile, options=options, service_log_path=geckodriverLogFile) File "/usr/local/lib/python3.8/site-packages/selenium/webdriver/firefox/webdriver.py", line 164, in __init__ self.service.start() File "/usr/local/lib/python3.8/site-packages/selenium/webdriver/common/service.py", line 81, in start raise WebDriverException( selenium.common.exceptions.WebDriverException: Message: 'geckodriver' executable needs to be in PATH.`
**No such file or directory: '/config/drivers/geckodriver'** --> pourtant bien présent avec les droits

3. En console SSH, si j'essai de lancer geckodriver, j'obtiens : 
`bash: geckodriver: command not found`
donc fichier bien présent mais il n'arrive pas à l'exécuter ?

4. Dans HomeAssistant OS, l'utilisation de Docker et des containers ne pourrait-elle pas être à l'origine de ces incompatibilités ? j'avoue que ça me dépasse complètement...

Merci encore pour ton aide :)

**Comment by ssenart** [comment]
Je me suis installé un HassIO sur Raspberry Pi 3 avec l'image hassos_rpi3-64-5.10.img.xz.
J'ai reproduit ton problème immédiatement.

Cela vient que l'image embarque une version 64 bits de Alpine Linux (architecture aarch64) et que l'on essaie d'utiliser un binaire 32 bits du geckodriver (architecture arm7hf).
Peut-être qu'il est possible de trouver des librairies permettant la compatibilité. Je n'ai pas cherché.

Par contre, j'ai recompilé le geckodriver en aarch64.

J'ai tout expliqué ici : https://github.com/ssenart/ha-custom_components/blob/master

En suivant ces explications, j'ai réussi à récupérer mes données Gazpar dans HassIO.

Il y a tout le détail pour que tu y arrives aussi.

Tiens-moi au courant !

**Comment by yvavasseur** [comment]
Génial, ça fonctionne, merci infiniment pour tout le temps que tu as consacré à mon problème ! :) 
👍

**Reply by cedsib** [reply]
Est-ce que @yvavasseur tu pourrais nous faire un mini tuto pour que ceux qui semblent un chouilla moins experts que toi mais qui ont très envie de récupérer ces données là dans HA (même instal que toi sur un RPI)

**Reply by ssenart** [reply]
Salut cedsib, tu peux trouver des indications ici : https://github.com/ssenart/ha-custom_components

Surtout sur les spécificités HassIO.

**Reply by cedsib** [reply]
Merci ! en effet y'a un peu de taf.

**Comment by ssenart** [comment]
Merci pour ton retour. Je suis content que ça fonctionne !
